src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -274,19 +274,41 @@ internal NotRecognizedSemaphore(int initialCount, JoinableTaskContext joinableTa
             {
                 Requires.NotNull(operation, nameof(operation));
 
+                // Note: this code is duplicated and not extracted to minimize allocating extra async state machines.
+                // For performance reasons, we want to minimize the number of Joins performed, and also keep the size of the
+                // JoinableCollection to a minimum. This also means awaiting on the semaphore outside of a JTF.RunAsync. This requires
+                // us to not ConfigureAwait(true) on the semaphore. However, that breaks our contract of resuming on the sync context
+                // you started on. To partially fix this, we will at least resume you on the main thread or thread pool.
+                AsyncSemaphore.Releaser releaser = default;
+                bool resumeOnMainThread = this.joinableTaskCollection.Context.IsOnMainThread;
+                using (this.joinableTaskCollection?.Join())
+                {
+                    // Use ConfiguredAwaitRunInline() as ConfigureAwait(true) will deadlock due to not being inside a JTF.RunAsync().
+                    releaser = await this.semaphore.EnterAsync(cancellationToken).ConfigureAwaitRunInline();
+                }
+
                 await this.ExecuteCoreAsync(async delegate
                 {
-                    using (this.joinableTaskCollection?.Join())
+                    if (resumeOnMainThread)
                     {
-                        var releaser = await this.semaphore.EnterAsync(cancellationToken).ConfigureAwait(true);
-                        try
-                        {
-                            await operation().ConfigureAwaitRunInline();
-                        }
-                        finally
-                        {
-                            DisposeReleaserNoThrow(releaser);
-                        }
+                        // Return to the main thread if we started there.
+                        await this.joinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
>() [](start = 78, length = 2)

We should pass in `cancellationToken`.
We should also be prepared to release the semaphore if it were to throw. So I believe this whole if/else block should be moved into the try below.